### PR TITLE
lstm: Fix issues reported by Coverity scan

### PR DESCRIPTION
**Pull request by stweil**
Signed-off-by: Stefan Weil <sw@weilnetz.de>

### CODE REVIEW

File: lstm/lstmtrainer.cpp [hunk 1]
```diff
@@ -149,7 +149,7 @@ void LSTMTrainer::InitCharSet(const UNICHARSET& unicharset,
 // ctc_mode controls how the truth text is mapped to the network targets.
 // Note: Call before InitNetwork!
 void LSTMTrainer::InitCharSet(const UNICHARSET& unicharset,
-                              const UnicharCompress recoder) {
+                              const UnicharCompress& recoder) {
```
Comment (amitdo) [inline lstm/lstmtrainer.cpp]: `UnicharCompress` can be `NULL`.

In the same file you have this:
`LSTMTrainer::EncodeString(const STRING& str, const UNICHARSET& unicharset, ...`
and
`if (recoder != NULL) {`

**Comment by stweil** [comment]
That's a different function which either gets called with `&recoder_` or with `NULL` as value for parameter `recoder`. It is unrelated to my modification.

**Comment by amitdo** [comment]
My point is that recoder may be NULL and you should use a pointer instead of a reference.

**Comment by stweil** [comment]
I don't think so. The old code passed an `UnicharCompress` object. Passing a reference is the optimized equivalent. How would you assign a NULL pointer to a `UnicharCompress` object (not a pointer) as it is done in the function?

**Comment by amitdo** [comment]
I will recheck it after a good sleep :)

**Comment by amitdo** [comment]
Stefan,
I tried to examine the call hierarchy, but I didn't find a call to this variant of `InitCharSet` overloaded method.

**Comment by amitdo** [comment]
@stweil?

**Comment by stweil** [comment]
Your analysis is correct: one of the `InitCharSet` methods is unused and can be removed.

@theraysmith, are there plans to use that interface in the future, or can `void InitCharSet(const UNICHARSET& unicharset, const UnicharCompress recoder)` be removed?

@zdenop, I think this PR can be applied anyway – we can also remove a fixed method in a later PR.

**Comment by theraysmith** [comment]
The alternative variant of InitCharset is used internally at Google to save accessing the script-dir directory from a trainer running in a Google data center. It is easier to pack the unicharset and recoder in a proto buffer in a separate step. If anything the other variant could prepare the recoder and then call the one that takes a recoder.

**Comment by stweil** [comment]
I rebased the PR now to fix a merge conflict in `lstm/lstmtrainer.cpp`.

**Comment by stweil** [comment]
Continuous integration fails because of missing `tesseract::LSTMTester::LSTMTester(long)`. This is unrelated to this PR.

**Comment by stweil** [comment]
I rebased to fix the build with CMake in the continuous integration.

**Comment by amitdo** [comment]
TIP: You can also close the PR and reopen it a the CI builds will rerun.

**Comment by stweil** [comment]
Thanks, I knew that. In this case the trick would not have worked because the CI problem was only fixed in git master.

**Comment by amitdo** [comment]
You are right :)

**Pull request merged**